server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -522,15 +524,28 @@ private static void throwOnCopyToOnObject(Mapper mapper, List<String> copyToFiel
         );
     }
 
-    private static void parseObject(final DocumentParserContext context, String currentFieldName) throws IOException {
+    private static int parseObject(final DocumentParserContext context, String currentFieldName, long arrayObjectsLimit, int alreadyCounted)
+        throws IOException {
+        int newCount = 0;
         assert currentFieldName != null;
         context.setImmediateXContentParent(context.parser().currentToken());
         Mapper objectMapper = context.getMapper(currentFieldName);
         if (objectMapper != null) {
             doParseObject(context, currentFieldName, objectMapper);
         } else {
+            newCount = alreadyCounted + 1;
+            if (newCount > arrayObjectsLimit) {
```

Comment:
Clarifying question (this code is a bit difficult to follow and I have not looked at it for a while): do we know for sure that the field being requested is always the same? It's  a bit weird to carry around the counter, as it needs to be incremented only in the else branch.